### PR TITLE
Regex handling in columns

### PR DESCRIPTION
**Pull request by bkamins**
@nalimilan - this follows our discussion on Slack. If you find the changes OK I will update the documentation accordingly and write tests.

### CODE REVIEW

File: src/abstractdataframe/abstractdataframe.jl [hunk 1]
```diff
@@ -608,6 +608,9 @@ end
 completecases(df::AbstractDataFrame, cols::AbstractVector) =
     completecases(df[cols])
 
+completecases(df::AbstractDataFrame, cols::Regex) =
```
Comment (nalimilan) [inline src/abstractdataframe/abstractdataframe.jl]: Could just use `::Union` in the previous method.

File: src/dataframe/dataframe.jl [hunk 2]
```diff
@@ -303,6 +303,8 @@ function Base.getindex(df::DataFrame, col_inds::AbstractVector)
     return DataFrame(new_columns, Index(_names(df)[selected_columns]))
 end
 
+Base.getindex(df::DataFrame, col_inds::Regex) =getindex(df, index(df)[col_inds])
```
Comment (nalimilan) [inline src/dataframe/dataframe.jl]:
```suggestion
Base.getindex(df::DataFrame, col_inds::Regex) = getindex(df, index(df)[col_inds])
```

File: src/dataframe/dataframe.jl [hunk 3]
```diff
@@ -512,6 +517,10 @@ function Base.setindex!(df::DataFrame, val::Any, col_inds::AbstractVector{<:Colu
     return df
 end
 
+# df[Regex] = value
```
Comment (nalimilan) [inline src/dataframe/dataframe.jl]: The existing code says "Single Item" rather than "value".

File: src/dataframe/dataframe.jl [hunk 4]
```diff
@@ -1191,6 +1204,13 @@ function disallowmissing!(df::DataFrame, cols::AbstractVector{Bool})
     df
 end
 
+function disallowmissing!(df::DataFrame, cols::Regex)
```
Comment (nalimilan) [inline src/dataframe/dataframe.jl]: `disallowmissing!(df, index(df)[cols])` would be shorter.

File: src/dataframe/dataframe.jl [hunk 5]
```diff
@@ -1277,6 +1297,13 @@ function categorical!(df::DataFrame, cnames::Vector{<:Union{Integer, Symbol}};
     df
 end
 
+function categorical!(df::DataFrame, cnames::Regex; compress::Bool=false)
```
Comment (nalimilan) [inline src/dataframe/dataframe.jl]: Same remark as for `disallowmissing!`.

File: src/abstractdataframe/abstractdataframe.jl (hunk 1 above)
Comment (bkamins) [reply src/abstractdataframe/abstractdataframe.jl]: fixed

File: src/dataframe/dataframe.jl (hunk 3 above)
Comment (bkamins) [reply src/dataframe/dataframe.jl]: This is exactly the point - I have written `value` on purpose. This method is catch-all for any `v` and dispatches accordingly.

Using this approach we "inherit" what is allowed from other `setindex!` definitions. When we remove implicit broadcasting in the other methods it will automatically disallowed for `Regex` specification of columns.

Comment (bkamins) [reply src/dataframe/dataframe.jl]: I do the same in line https://github.com/JuliaData/DataFrames.jl/pull/1819/files#diff-0aa1ff52d56badabf7d1511d3dc3f279R684 for `setindex!` where row range is passed.

File: src/dataframe/dataframe.jl (hunk 4 above)
Comment (bkamins) [reply src/dataframe/dataframe.jl]: fixed

File: src/dataframe/dataframe.jl (hunk 5 above)
Comment (bkamins) [reply src/dataframe/dataframe.jl]: fixed

**Comment by bkamins** [comment]
>  I wonder whether we'd better avoid adding some setindex! methods which perform implicit broadcasting, since we want to deprecate them soon.

As I have written in the comment the methods for `setindex!` only pass functionality to other `setindex!` methods so I think it is OK.

Also I hope that when we have final decision on https://github.com/JuliaData/DataFrames.jl/pull/1804 it can be merged very soon and then `setindex!` changes PR will be opened soon also. So all should be merged before the next release anyway.

File: src/other/index.jl [hunk 6]
```diff
@@ -204,6 +204,9 @@ end
                         "DataFrame only supports indexing columns with integers, symbols or boolean vectors"))
 end
 
+@inline function Base.getindex(x::AbstractIndex, rx::Regex)
+    getindex(x, filter(name -> occursin(rx, String(name)), _names(x)))
+end
 # Helpers
```
Comment (nalimilan) [inline src/other/index.jl]: Add empty line.

Comment (bkamins) [reply src/other/index.jl]: Added.

This PR is not ready for a review yet - I will let you know (and remove WIP) when its ready (I asked for your initial comments to make sure which way we go forward as there is a lot of tests to write). Thank you.

**Comment by bkamins** [comment]
This PR should be good to have a look at.

**Comment by bkamins** [comment]
It would be also great to have a comment on this as after I merge https://github.com/JuliaData/DataFrames.jl/pull/1804 it would be good to have it in place before further `setindex!` work.

File: test/dataframerow.jl [hunk 7]
```diff
@@ -42,27 +47,41 @@ end
 
     r = DataFrameRow(df, 2, :)
     @test r[:] === r
```
Comment (nalimilan) [inline test/dataframerow.jl]: BTW, shouldn't this return a named tuple?

File: test/dataframerow.jl [hunk 8]
```diff
@@ -42,27 +47,41 @@ end
...
+    @test isequal(r[r""], r)
```
Comment (nalimilan) [inline test/dataframerow.jl]: We use `≈` consistently I think. Same below.

File: test/dataframerow.jl [hunk 9]
```diff
@@ -42,27 +47,41 @@ end
...
     @test view(r, :) === r
+    @test isequal(view(r, r""), r)
     @test r[3] == "B"
+    @test r[r"c"] == r[[3]]
     @test_throws BoundsError r[5]
     @test view(r, 3)[] == "B"
+    @test view(r, r"c")[1] == "B"
+    view(r, r"c") == view(r, [3])
```
Comment (nalimilan) [inline test/dataframerow.jl]:
```suggestion
    @test view(r, r"c") == view(r, [3])
```

File: test/dataframerow.jl [hunk 10]
```diff
@@ -42,27 +47,41 @@ end
...
     view(r, 3)[] = "BB"
     @test df.c[2] == "BB"
     @test_throws MethodError r[true]
     @test_throws MethodError view(r, true)
     @test copy(r[[:c,:a]]) == (c = "BB", a = 2)
+    @test copy(r[r"[ac]"]) == (a = 2, c = "BB")
+    @test copy(r[r"x"]) == NamedTuple()
```
Comment (nalimilan) [inline test/dataframerow.jl]: Should this also be tested for an empty vector of symbols? Same below.

File: test/dataframerow.jl [hunk 11]
```diff
@@ -71,8 +90,22 @@ end
     @test copy(view(r, [:c,:b])) == (c = "CC", b = 1.2)
     @test copy(view(r, [:c,:b])) == (c = "CC", b = 1.2)
     @test copy(r[[false, true]]) == (c = "CC",)
+    @test copy(r[r"b"]) == (b = 1.2,)
```
Comment (nalimilan) [inline test/dataframerow.jl]: Why not test `r"[cb]"` for consistency with the test above?

File: test/dataframerow.jl [hunk 12]
```diff
@@ -71,8 +90,22 @@ end
...
+    @test copy(view(r, r"b")) == (b = 1.2,)
     r.c = "C"
     @test df.c[3] == "C"
+
+    df = DataFrame([1 2 3 4
+                    5 6 7 8])
+    r = df[1, r"[1-3]"]
+    @test names(r) == [:x1, :x2, :x3]
+    r[:] = 10
+    @test Matrix(df) == [10 10 10 4
```
Comment (nalimilan) [inline test/dataframerow.jl]: It would be more logical to do this:
```julia
    @test df == DataFrame([10 10 10 4
                           5  6  7 8])
```
Same below.

File: test/dataframerow.jl [hunk 13]
```diff
@@ -230,12 +263,39 @@ end
     df = deepcopy(ref_df)[1:3]
 
     @test parent(df[2, :]) === df
+    @test parentindices(df[2,[]]) == (2, Int[])
```
Comment (nalimilan) [inline test/dataframerow.jl]:
```suggestion
    @test parentindices(df[2, []]) == (2, Int[])
```

File: test/index.jl [hunk 14]
```diff
@@ -178,4 +178,53 @@ end
...
 
+@testset "Regex indexing" begin
+    i = Index()
+    push!(i, :x1)
+    push!(i, :x12)
+    push!(i, :x131)
+    push!(i, :y13)
+    push!(i, :yy13)
+    @test i[r"x1."] == [2, 3]
+    @test i[r"xx"] == Int[]
```
Comment (nalimilan) [inline test/index.jl]:
```suggestion
    @test i[r"xx"] == []
```

It's a bit misleading since the type isn't checked. Same below.

File: test/indexing.jl [hunk 15]
```diff
@@ -8,34 +8,44 @@ using Test, DataFrames
...
     @test df[1:2] == DataFrame(a=1:3, b=4:6)
+    @test df[r"[ab]"] == DataFrame(a=1:3, b=4:6)
     @test eachcol(df, false)[1] === df[1]
     @test eachcol(view(df,1:2), false)[1] == eachcol(df, false)[1]
     @test eachcol(df[1:2], false)[1] == eachcol(df, false)[1]
+    @test eachcol(df[r"[ab]"], false)[1] == eachcol(df, false)[1]
     @test eachcol(df[1:2], false)[1] !== eachcol(df, false)[1]
     @test df[:] == df
+    @test df[r""] == df
     @test df[:] !== df
```
Comment (nalimilan) [inline test/indexing.jl]: Also test this for regexes? Same for other things in this test set.

File: test/indexing.jl [hunk 16]
```diff
@@ -55,33 +67,52 @@ end
     @test view(df, 1) isa SubArray
     @test view(df, 1:2) isa SubDataFrame
     @test view(df, 1:2) == df[1:2]
+    @test view(df, r"[ab]") isa SubDataFrame
+    @test view(df, r"[ab]") == df[1:2]
     @test view(df, :) isa SubDataFrame
     @test view(df, :) == df
-    @test parent(view(df, :)) === df
```
Comment (nalimilan) [inline test/indexing.jl]: Keep this?

File: docs/src/lib/indexing.md [hunk 17]
```diff
@@ -19,9 +19,10 @@ The rules for a valid type of index into a column are the following:
     * an `Integer` that is not `Bool`;
 * a vector, later denoted as `cols`:
     * a vector of `Symbol` (does not have to be a subtype of `AbstractVector{Symbol}`);
-    * a vector of `Integer` other than `Bool` (does not have to be a subtype of `AbstractVector{<:Integer}`);
-    * a vector of `Bool` that has to be a subtype of `AbstractVector{Bool}`.
-    * a colon.
+    * a vector of `Integer` other than `Bool` (does not have to be a subtype of `AbstractVector{<:Integer}`).
```
Comment (nalimilan) [inline docs/src/lib/indexing.md]:
```suggestion
    * a vector of `Integer` other than `Bool` (does not have to be a subtype of `AbstractVector{<:Integer}`);
```

File: docs/src/lib/indexing.md [hunk 18]
```diff
@@ -19,9 +19,10 @@ The rules for a valid type of index into a column are the following:
...
+    * a vector of `Bool` that has to be a subtype of `AbstractVector{Bool}`;
+    * a colon;
+    * a regular expression, which gets expanded to a vector of matching column indices.
```
Comment (nalimilan) [inline docs/src/lib/indexing.md]:
```suggestion
    * a regular expression, which gets expanded to a vector of matching column names.
```

File: src/abstractdataframe/abstractdataframe.jl [hunk 19]
```diff
@@ -868,6 +868,9 @@ nonunique(df, 1)
 
 """
 function nonunique(df::AbstractDataFrame)
+    if ncol(df) == 0
+        throw(ArgumentError("finding duplicate rows in zero column data frame is not allowed"))
```
Comment (nalimilan) [inline src/abstractdataframe/abstractdataframe.jl]:
```suggestion
        throw(ArgumentError("finding duplicate rows in data frame with no columns is not allowed"))
```

File: src/abstractdataframe/abstractdataframe.jl [hunk 20]
```diff
@@ -884,13 +887,17 @@ nonunique(df::AbstractDataFrame, cols::Any) = nonunique(df[cols])
 Base.unique!(df::AbstractDataFrame) = deleterows!(df, findall(nonunique(df)))
 Base.unique!(df::AbstractDataFrame, cols::AbstractVector) =
     deleterows!(df, findall(nonunique(df, cols)))
+Base.unique!(df::AbstractDataFrame, cols::Regex) =
```
Comment (nalimilan) [inline src/abstractdataframe/abstractdataframe.jl]: Can be merged with previous and next method.

File: src/abstractdataframe/abstractdataframe.jl [hunk 21]
```diff
@@ -884,13 +887,17 @@ nonunique(df::AbstractDataFrame, cols::Any) = nonunique(df[cols])
...
+    deleterows!(df, findall(nonunique(df, cols)))
 Base.unique!(df::AbstractDataFrame, cols::Union{Integer, Symbol, Colon}) =
     deleterows!(df, findall(nonunique(df, cols)))
 
 # Unique rows of an AbstractDataFrame.
 Base.unique(df::AbstractDataFrame) = df[(!).(nonunique(df)), :]
 Base.unique(df::AbstractDataFrame, cols::AbstractVector) =
     df[(!).(nonunique(df, cols)), :]
+Base.unique(df::AbstractDataFrame, cols::Regex) =
```
Comment (nalimilan) [inline src/abstractdataframe/abstractdataframe.jl]: Same.

File: src/abstractdataframe/reshape.jl [hunk 22]
```diff
@@ -258,7 +258,7 @@ function _unstack(df::AbstractDataFrame, rowkey::Int,
 end
 
 unstack(df::AbstractDataFrame, rowkey::ColumnIndex,
-        colkey::ColumnIndex, value::ColumnIndex) =
+        colkey::Int, value::Int) =
```
Comment (nalimilan) [inline src/abstractdataframe/reshape.jl]: Why do `index(df)[rowkey]` if we know the index is an integer already?

File: src/dataframe/dataframe.jl [hunk 23]
```diff
@@ -303,6 +303,8 @@ function Base.getindex(df::DataFrame, col_inds::AbstractVector)
     return DataFrame(new_columns, Index(_names(df)[selected_columns]))
 end
 
+Base.getindex(df::AbstractDataFrame, col_inds::Regex) = getindex(df, index(df)[col_inds])
```
Comment (nalimilan) [inline src/dataframe/dataframe.jl]: Why not merge this with previous method? Same below.

File: src/dataframerow/dataframerow.jl [hunk 24]
```diff
@@ -78,16 +78,18 @@ Base.@propagate_inbounds Base.view(adf::AbstractDataFrame, rowind::Integer, ::Co
 Base.@propagate_inbounds Base.view(adf::AbstractDataFrame, rowind::Integer,
                                    colinds::AbstractVector) =
     DataFrameRow(adf, rowind, colinds)
+Base.@propagate_inbounds Base.view(adf::AbstractDataFrame, rowind::Integer,
```
Comment (nalimilan) [inline src/dataframerow/dataframerow.jl]: Merge with previous method.

File: src/other/index.jl [hunk 25]
```diff
@@ -201,7 +201,11 @@ end
     end
     idxs[1] isa Symbol && return getindex(x, convert(Vector{Symbol}, idxs))
     throw(ArgumentError("idxs[1] has type $(typeof(idxs[1])); "*
-                        "DataFrame only supports indexing columns with integers, symbols or boolean vectors"))
+                        "DataFrame only supports indexing columns with vectors of integers or symbols"))
```
Comment (nalimilan) [inline src/other/index.jl]: This could be confusing, as we allow indexing data fames with integers, symbols, regexes, and vectors of integers, symbols or Booleans.

File: src/subdataframe/subdataframe.jl [hunk 26]
```diff
@@ -102,6 +102,8 @@ Base.@propagate_inbounds Base.getindex(sdf::SubDataFrame, colind::ColumnIndex) =
     view(parent(sdf), rows(sdf), parentcols(index(sdf), colind))
 Base.@propagate_inbounds Base.getindex(sdf::SubDataFrame, colinds::AbstractVector) =
     SubDataFrame(parent(sdf), rows(sdf), parentcols(index(sdf), colinds))
+Base.@propagate_inbounds Base.getindex(sdf::SubDataFrame, colinds::Regex) =
```
Comment (nalimilan) [inline src/subdataframe/subdataframe.jl]: Merge with previous method. Same below.

File: test/dataframerow.jl (hunk 7 above)
Comment (bkamins) [reply test/dataframerow.jl]: Well - we have that `r[cols]` is a `DataFrameRow`. We could change the rules so that:
* `dfr[cols]` is a `NamedTuple`
* `@view dfr[cols]` is a `DataFrameRow`

This is a separate PR I think. Do you think we should change this? I am not really sure what is best, and in general I would leave it as is, as it is not a top priority operation anyway. Returning `NamedTuple` from `dfr[cols]` is kind of inconsistent with the fact that we allow `dfr[cols] = something`.

File: src/abstractdataframe/abstractdataframe.jl (hunk 20 above)
Comment (bkamins) [reply src/abstractdataframe/abstractdataframe.jl]: Merging them would lead to method ambiguity with methods defined for `unique!` in Base. The problematic method there is `unique!(f, A::AbstractArray{T,1} where T)`.

File: src/dataframe/dataframe.jl (hunk 23 above)
Comment (bkamins) [reply src/dataframe/dataframe.jl]: This would lead to method ambiguities.

File: test/dataframerow.jl (hunk 7 above)
Comment (nalimilan) [reply test/dataframerow.jl]: Scratch that, for some reason I thought `dfr[cols]` returns a named tuple.

File: test/indexing.jl (hunk 16 above)
Comment (bkamins) [reply test/indexing.jl]: ok

File: src/abstractdataframe/reshape.jl (hunk 22 above)
Comment (bkamins) [reply src/abstractdataframe/reshape.jl]: `rowkey` can be a `Symbol`, but we can drop it for `colkey` and `value`. Fixed.

File: src/other/index.jl (hunk 25 above)
Comment (bkamins) [reply src/other/index.jl]: fixed.

File: test/dataframerow.jl (hunk 8 above)
Comment (bkamins) [reply test/dataframerow.jl]: fixed

File: test/dataframerow.jl (hunk 10 above)
Comment (bkamins) [reply test/dataframerow.jl]: fixed

File: test/dataframerow.jl (hunk 11 above)
Comment (bkamins) [reply test/dataframerow.jl]: added.

File: test/dataframerow.jl (hunk 12 above)
Comment (bkamins) [reply test/dataframerow.jl]: fixed

File: test/indexing.jl (hunk 15 above)
Comment (bkamins) [reply test/indexing.jl]: fixed

File: src/abstractdataframe/abstractdataframe.jl (hunk 21 above)
Comment (bkamins) [reply src/abstractdataframe/abstractdataframe.jl]: merged

File: src/dataframerow/dataframerow.jl (hunk 24 above)
Comment (bkamins) [reply src/dataframerow/dataframerow.jl]: done

File: src/subdataframe/subdataframe.jl (hunk 26 above)
Comment (bkamins) [reply src/subdataframe/subdataframe.jl]: regex

**Comment by bkamins** [comment]
I have changed the method layout a bit and it was possible to merge most of them. It should be good now to have a look at.

**Comment by bkamins** [comment]
After this PR is merged the plan is to:
* implement `Not` and deprecate `deletecols` and `deletecols!` in the next PR
* then to cleanup `setindex!` in yet next PR

when these things are done and https://github.com/JuliaData/DataFrames.jl/pull/1840 is finished we shall be ready for a release I think.
